[Quant][fx] Add lowering for functional conv

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73708

Summary: This adds functionality to lower reference models
involving functional conv in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_conv

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D34648870](https://our.internmc.facebook.com/intern/diff/D34648870)

andrewor14 committed:
[Quant][fx] Add lowering for functional conv

Summary: This adds functionality to lower reference models
involving functional conv in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_conv

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -37,6 +37,12 @@
     torch.nn.functional.instance_norm : [4],
 }
 
+CONV_FUNCTIONAL_OPS = {
```
let's just define this twice, since quantization_patterns.py will soon be removed and after that we only need this in lowering

andrewor14 replied:
I see, I had actually defined it twice before but refactored it. Ok

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -4035,6 +4035,9 @@ def forward(self, x):
             (True, False),  # functional relu
         )
         for dim, quant_type, use_bias, has_relu, f_relu in options:
+            # TODO(andrew): fix bias case
```
are we going to support bias in the same PR? or is it already supported?

jerryzh168 replied:
looks like it is already handled

andrewor14 replied:
It currently fails the test with "tensor values not close enough". I am investigating and will push a fix within this PR.

jerryzh168 replied:
why? aren't we producing the same model in the end?

jerryzh168 replied:
or are you comparing reference with non-reference again? which we didn't do before?

andrewor14 replied:
It was because both models with and without bias matched the same pattern, but the lowering code assumed the model doesn't have a bias in both cases. This is fixed in the latest commit which matches the patterns manually and makes no such assumptions

jerryzh168 reviewed (approved):
looks great, thanks for putting this up quickly!

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional conv"


Summary: This adds functionality to lower reference models
involving functional conv in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_conv

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional conv"


Summary: This adds functionality to lower reference models
involving functional conv in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_conv

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34648870).

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -371,77 +380,61 @@ def _lower_weighted_ref_functional(
+        # Step 0: Find nodes that match this pattern (dequantize - functional op - quantize)
+        # We search for the pattern backwards, starting with the quantize node
+        # Quantize node args: (func, scale, zp, dtype)
+        if n.target != torch.quantize_per_tensor:
```
we should probably do
```
if n.op != "call_function" or n.target != torch.quantize_per_tensor:
    continue
```

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -371,77 +380,61 @@ def _lower_weighted_ref_functional(
+            func_node = relu_node.args[0]
+        else:
+            relu_node = None
+        if func_node.target not in LOWER_FUNCTIONAL_MAP:
```
will need to make sure func_node.op == "call_function" here as well

andrewor14 replied:
Actually I check for that in L393, but I can merge the two checks

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -371,77 +380,61 @@ def _lower_weighted_ref_functional(
+        # Step 1: Replace quantized weights with packed weights, which will be folded later
+        # Linear args: (dequantized inputs, dequantized weights[, bias])
+        # Conv args: (dequantized inputs, dequantized weights[, bias, stride, padding, dilation, groups])
+        (input_dq_node, weight_dq_node, *remaining_func_args) = func_node.args
```
do we need to verify `weight_dq_node` is a dequant node as well?

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -371,77 +380,61 @@ def _lower_weighted_ref_functional(
+        # Linear args: (dequantized inputs, dequantized weights[, bias])
+        # Conv args: (dequantized inputs, dequantized weights[, bias, stride, padding, dilation, groups])
+        (input_dq_node, weight_dq_node, *remaining_func_args) = func_node.args
+        quantized_weight = weight_dq_node.args[0]
```
and quantize_weight is a quantize_per_tensor node?

andrewor14 replied:
Actually it can be a `quantize_per_tensor` or `quantize_per_channel`, but yes I will add the check

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -371,77 +380,61 @@ def _lower_weighted_ref_functional(
+        func_node.target = q_relu_func if relu_node is not None else q_func
+        func_node.args = (input_dq_node.args[0], packed_weight, output_scale_node, output_zp_node)
+        q_node.replace_all_uses_with(func_node)
+        output_zp_node.append(func_node)
```
what does this mean?

andrewor14 replied:
This means `func_node` is moved to after `output_zp_node` (actually this line was here before already)

jerryzh168 replied:
maybe add a comment? `append` has the same name as list.append, which is pretty confusing

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional conv"


Summary: This adds functionality to lower reference models
involving functional conv in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_conv

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D34648870](https://our.internmc.facebook.com/intern/diff/D34648870)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34648870).

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional conv"


Summary: This adds functionality to lower reference models
involving functional conv in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_conv

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D34648870](https://our.internmc.facebook.com/intern/diff/D34648870)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34648870).

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional conv"


Summary: This adds functionality to lower reference models
involving functional conv in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_conv

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D34648870](https://our.internmc.facebook.com/intern/diff/D34648870)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34648870).

andrewor14 committed:
Update on "[Quant][fx] Add lowering for functional conv"


Summary: This adds functionality to lower reference models
involving functional conv in FX.

Test Plan: python test/test_quantization.py TestQuantizeFxOps.test_functional_conv

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D34648870](https://our.internmc.facebook.com/intern/diff/D34648870)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34648870).

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -371,77 +380,74 @@ def _lower_weighted_ref_functional(
+        if func_node.op != "call_function" or func_node.target not in LOWER_FUNCTIONAL_MAP:
+            continue
+        (input_dq_node, weight_dq_node, *remaining_func_args) = func_node.args
+        if input_dq_node.target != "dequantize" or weight_dq_node.target != "dequantize":
```
in general we should check for both op and target, op for dequantize is "call_method"
